src/firmware/boards/legacy_robot_stm32f4/io/receive.c

```diff
@@ -191,107 +196,79 @@ void receive_tick(log_record_t *record)
-}
-
-void handle_drive_packet(uint8_t *dma_buffer)
+void handle_drive_packet(uint8_t *packet_data, size_t packet_size)
 {
-    // Grab emergency stop status from the end of the frame.
-    bool estop_run =
-        !!dma_buffer[MESSAGE_PAYLOAD_ADDR + NUM_ROBOTS * RECEIVE_DRIVE_BYTES_PER_ROBOT];
-
-    // Grab a pointer to the robot’s own data block.
-    const uint8_t *robot_data =
-        dma_buffer + MESSAGE_PAYLOAD_ADDR + RECEIVE_DRIVE_BYTES_PER_ROBOT * robot_index;
+    bool estop_triggered                    = packet_data[0] != 1;
+    bool feedback_requested_from_this_robot = packet_data[1] == robot_index;
+    uint8_t packet_robot_id                 = packet_data[2];
 
     // Check if feedback should be sent.
-    if (robot_data[0] & 0x80)
+    if (feedback_requested_from_this_robot)
     {
         feedback_pend_normal();
     }
 
-    // Extract the serial number.
-    uint8_t serial = *robot_data++ & 0x0F;
-
-    // Construct the individual 16-bit words sent from the host.
-    uint16_t words[4U];
-    for (unsigned int i = 0U; i < 4U; ++i)
+    // Check if this drive packet was intended for us, if not we can stop here
+    if (packet_robot_id != robot_index)
     {
-        words[i] = *robot_data++;
-        words[i] |= (uint16_t)*robot_data++ << 8;
+        return;
     }
 
-    // In case of emergency stop, treat everything as zero
-    // except the chicker discharge bit (that can keep its
-    // status).
-    if (!estop_run)
+    // Reset timeout.
+    timeout_ticks = 1000U / portTICK_PERIOD_MS;
+
+    // Figure out what primitive to run
+    primitive_params_t pparams = {0};
+    unsigned int primitive     = 0;
+    if (estop_triggered)
+    {
+        // Set the primitive to be a stop primitive
+        primitive         = 0;
+        pparams.params[0] = 0;
+        pparams.params[1] = 0;
+        pparams.params[2] = 0;
+        pparams.params[3] = 0;
+        pparams.slow      = true;
+        pparams.extra     = 0;
+
+        // Disable charging and discharge through chicker
+        charger_enable(false);
+        chicker_discharge(true);
+    }
+    else
     {
-        static const uint16_t MASK[4] = {0x0000, 0x4000, 0x0000, 0x0000};
-        for (unsigned int i = 0; i != 4; ++i)
+        // Decode the primitive
+        pb_istream_t pb_in_stream = pb_istream_from_buffer(packet_data, packet_size - 3);
+        PrimitiveMsg prim_msg     = PrimitiveMsg_init_zero;
+        if (!pb_decode(&pb_in_stream, PrimitiveMsg_fields, &prim_msg))
         {
-            words[i] &= MASK[i];
+            // If we failed to decode the message, it's likely malformed, so we should not
+            // proceed
+            return;
         }
+
+        pparams.slow      = prim_msg.slow;
+        pparams.extra     = prim_msg.extra_bits;
+        pparams.params[0] = prim_msg.parameter1;
+        pparams.params[1] = prim_msg.parameter2;
+        pparams.params[2] = prim_msg.parameter3;
+        pparams.params[3] = prim_msg.parameter3;
```

Comment:
something funky going on with indexes? 1,2,3,3?

Reply:
annnnnd that's a bug! good catch, love code without tests.